File: neetcode-150/java/Stack/isValid.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> brackets = new Stack<>();
        Map<Character, Character> bracketLookup = new HashMap<>(3);
        bracketLookup.put(')', '(');
        bracketLookup.put('}', '{');
        bracketLookup.put(']', '[');
        for (char c : s.toCharArray()) {
            if (bracketLookup.containsKey(c)) {
                if (brackets.empty() || !bracketLookup.get(c).equals(brackets.peek()))
                    return false;
                else
                    brackets.pop();
            } else {
                brackets.push(c);
            }
        }
        return brackets.empty();
    }
}

File: nexon/01-04-2025/isValid.py
class Solution:
    def isValid(self, s: str) -> bool:
        o, c = "({[", ")}]"
        st = []
        for i in s:
            if i in o:
                st.append(i)
            else:
                if st and st[-1] == o[c.index(i)]:
                    st.pop()
                else:
                    return False
        return len(st) == 0
        